kore/src/GraphTraversal.hs

```diff
@@ -0,0 +1,295 @@
+      Branch a (NonEmpty a)
+    | -- | no next state but not final (e.g., not goal state, or side
+      -- conditions do not hold)
+      Stuck a
+    | -- | final state (e.g., goal state reached, side conditions hold)
+      Final a
+    | -- | not stuck, but also not final (maximum depth reached before
+      -- finishing the proof)
+      Stopped a
+    | -- Future work:
+
+      -- | config matches a terminal pattern (or: is RHS of a
+      -- "terminal" rule) Needs to return that RHS
+      Terminal a
+    | -- | config matches a cut pattern ( aka: is LHS of a "cut" rule)
+      -- The respective RHS (or RHSs) are returned (if any)
+      Cut a [a]
+    deriving stock (Eq, Show, GHC.Generic)
+
+instance Functor TransitionResult where
+    fmap f = \case
+        StraightLine a -> StraightLine $ f a
+        Branch a as -> Branch (f a) $ NE.map f as
+        Stuck a -> Stuck $ f a
+        Final a -> Final $ f a
+        Stopped a -> Stopped $ f a
+        Terminal a -> Terminal $ f a
+        Cut a as -> Cut (f a) (map f as)
+
+instance Pretty a => Pretty (TransitionResult a) where
+    pretty = \case
+        StraightLine a -> single "StraightLine" a
+        Branch a as -> multi "Branch" "node" a "successors" (NE.toList as)
+        Stuck a -> single "Stuck" a
+        Final a -> single "Final" a
+        Stopped a -> single "Stopped" a
+        Terminal a -> single "Terminal" a
+        Cut a as -> multi "Cut" "node" a "successors" as
+      where
+        single :: Doc x -> a -> Doc x
+        single lbl a =
+            Pretty.vsep [lbl, Pretty.indent 4 $ Pretty.pretty a]
+
+        multi :: Doc x -> Doc x -> a -> Doc x -> [a] -> Doc x
+        multi lbl lbl1 a lbl2 as =
+            Pretty.vsep $
+                [ lbl
+                , Pretty.indent 2 $ "- " <> lbl1
+                , Pretty.indent 4 $ Pretty.pretty a
+                , Pretty.indent 2 $ "- " <> lbl2
+                ]
+                    <> map (Pretty.indent 4 . Pretty.pretty) as
+
+-- Graph traversal would always stop at Terminal/Cut, and _may_ stop
+-- at Branch, depending on configuration.
+
+isStuck, isFinal, isStopped, isTerminal, isCut, isBranch :: TransitionResult a -> Bool
+isStuck (Stuck _) = True
+isStuck _ = False
+isFinal (Final _) = True
+isFinal _ = False
+isStopped (Stopped _) = True
+isStopped _ = False
+isBranch (Branch _ _) = True
+isBranch _ = False
+isTerminal (Terminal _) = True
+isTerminal _ = False
+isCut (Cut _ _) = True
+isCut _ = False
+
+extractNext :: TransitionResult a -> [a]
+extractNext = \case
+    StraightLine a -> [a]
+    Branch _ as -> NE.toList as
+    Stuck _ -> []
+    Final _ -> []
+    Stopped _ -> []
+    Terminal _ -> []
+    Cut _ as -> as
+
+extractState :: TransitionResult a -> Maybe a
+extractState = \case
+    StraightLine _ -> Nothing
+    Branch a _ -> Just a
+    Stuck a -> Just a
+    Final a -> Just a
+    Stopped a -> Just a
+    Terminal a -> Just a
+    Cut a _ -> Just a
+
+type Step = [Prim]
```

Comment:
The former `Strategy prim` type can probably just be a list `[prim]`, we seem to only use `Seq` and `Apply`.